src/Features/Core/Portable/EmbeddedLanguages/Json/RoslynConvert.cs

```diff
@@ -0,0 +1,234 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.EmbeddedLanguages.Json;
+
+// Json.net uses Convert.ToInt64 when checking if numbers are legal (see
+// https://github.com/JamesNK/Newtonsoft.Json/blob/993215529562866719689206e27e413013d4439c/Src/Newtonsoft.Json/JsonTextReader.cs#L1926).
+// However, this is very expensive when it fails as it throws an exception (see
+// https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1871418).  As such, to avoid that, we have a copy of
+// Convert.ToInt64 here which uses `bool Try(out)` semantics instead of exception failure semantics.  Not ideal, but
+// necessary as long as the runtime exposes no other way to do this validation with the same semantics in a non-throwing
+// fashion.
+//
+// If this code is updated, it should follow the original BCL code as closely as possible, with the only changes being
+// to make it not throw, and that the result is returned in an out-position.  It should otherwise be kept the same.
+//
+// The normal json parser does not need this as we simply follow the json spec exactly wrt what is or is not legal with
+// numbers.  This code is only needed to match json.net's "quirks".
+
+/// <summary>
+/// Copy of helpers we need from https://github.com/dotnet/runtime/blob/2bfa26cebc917d05a3363078fa277ab5fee2651b/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L2178
+/// that do not throw.
+/// </summary>
+internal static class RoslynConvert
+{
+    public static bool TryToInt64(string? value, int fromBase, out long result)
+    {
+        if (fromBase != 2 && fromBase != 8 && fromBase != 10 && fromBase != 16)
+        {
+            result = 0;
+            return false;
+        }
+
+        if (value is null)
+        {
+            result = 0;
+            return true;
+        }
+
+        return RoslynParseNumbers.TryStringToLong(value.AsSpan(), fromBase, RoslynParseNumbers.IsTight, out result);
+    }
+}
+
+/// <summary>
+/// Copy of helpers we need from https://github.com/dotnet/runtime/blob/2bfa26cebc917d05a3363078fa277ab5fee2651b/src/libraries/System.Private.CoreLib/src/System/ParseNumbers.cs#L10
+/// that do not throw.
+/// </summary>
+internal static class RoslynParseNumbers
+{
+    private const int TreatAsUnsigned = 0x0200;
+    public const int IsTight = 0x1000;
+
+    public static unsafe bool TryStringToLong(ReadOnlySpan<char> s, int radix, int flags, out long result)
+    {
+        var pos = 0;
+        return TryStringToLong(s, radix, flags, ref pos, out result);
+    }
+
+    private static bool TryStringToLong(ReadOnlySpan<char> s, int radix, int flags, ref int currPos, out long finalResult)
+    {
+        finalResult = 0;
+        var i = currPos;
```

Comment:
the code that follows is a literal copy/paste, except for making it follow the try/out pattern instead of exceptions.  i have done nothing to validate the logic and i presume the runtime is doing the right thing here.

note: we have all our normal numeric json tests (and json.net tests) which maintain their smae behavior after making this change.  so we should be good.